### PR TITLE
TLVs draft PCEP 

### PR DESCRIPTION
**Pull request by Luis-Cepeda**

### CODE REVIEW

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyName.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                  Figure: The SRPOLICY-POL-NAME TLV format
+                  
+                  
+	 */
+
+	private String policyName;
+	
+	
+	public SRPolicyName(){
+		this.TLVType=56;
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyName.java:34]:
```suggestion
		this.TLVType=ObjectParameters.PCEP_TLV_SRPOLICY_POL_NAME;
```

Create the entry in ObjectParameters with the type

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathPreferenceTLV.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
+package es.tid.pce.pcep.objects.tlvs;
+
+import java.util.Objects;
+
+import es.tid.pce.pcep.objects.MalformedPCEPObjectException;
+import es.tid.protocol.commons.ByteHandler;
+
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathPreferenceTLV.java:7]: Add header with name of the TLV, reference (https://datatracker.ietf.org/doc/draft-ietf-pce-segment-routing-policy-cp/04/), and author

File: src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java [hunk 3]
```diff
@@ -214,17 +260,31 @@ public void decode() throws MalformedPCEPObjectException {
 			if (ObjectLength==16){
 				fin=true;
 			}
+			
 			offset=16;
 			while (!fin) {
 				int tlvtype=PCEPTLV.getType(this.getObject_bytes(), offset);
 				int tlvlength=PCEPTLV.getTotalTLVLength(this.getObject_bytes(), offset);
+				System.out.println("PRUEBASS:  " + tlvtype + "/ " +tlvlength);
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java:268]: remove log

File: src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java [hunk 4]
```diff
@@ -214,17 +260,31 @@ public void decode() throws MalformedPCEPObjectException {
...
 				int tlvlength=PCEPTLV.getTotalTLVLength(this.getObject_bytes(), offset);
+				System.out.println("PRUEBASS:  " + tlvtype + "/ " +tlvlength);
 				switch (tlvtype){
 				case ObjectParameters.PCEP_TLV_GLOBAL_ASSOCIATION_SOURCE:
 					global_association_source_tlv=new GlobalAssociationSourceTLV(this.getObject_bytes(), offset);
 					break;
 				case ObjectParameters.PCEP_TLV_EXTENDED_ASSOCIATION_ID:
 					extended_ssociation_id_tlv=new ExtendedAssociationIDTLV(this.getObject_bytes(), offset);
 					break;
+				case 56:
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java:276]: change with ObjectParameters.. XXX

File: src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java [hunk 5]
```diff
@@ -214,17 +260,31 @@ public void decode() throws MalformedPCEPObjectException {
...
+					sr_policy_name = new SRPolicyName(this.getObject_bytes(),offset);
+					break;
+				case 57:
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java:279]: change with ObjectParameters.. XXX

File: src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java [hunk 6]
```diff
@@ -214,17 +260,31 @@ public void decode() throws MalformedPCEPObjectException {
...
+					sr_policy_candidate_path_identifiers_tlv= new SRPolicyCandidatePathIdentifiersTLV(this.getObject_bytes(),offset);
+					break;
+				case 58:
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java:282]: change with ObjectParameters.. XXX

File: src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java [hunk 7]
```diff
@@ -214,17 +260,31 @@ public void decode() throws MalformedPCEPObjectException {
...
+					sr_policy_candidate_path_tlv=new SRPolicyCandidatePathNameTLV(this.getObject_bytes(),offset);
+					break;
+				case 59:
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java:285]: change with ObjectParameters.. XXX

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java [hunk 8]
```diff
@@ -0,0 +1,196 @@
+package es.tid.pce.pcep.objects.tlvs;
+
+import java.net.Inet4Address;
+import java.util.Objects;
+
+import es.tid.pce.pcep.objects.MalformedPCEPObjectException;
+import es.tid.protocol.commons.ByteHandler;
+import jdk.internal.org.jline.utils.Log;
+
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java:9]: add header with name, type, summary, reference and author

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+	private int reserved;
+	
+	private long originatorASN;
+	
+	private Inet4Address originatorAddress;
+	
+	private long discriminator;
+	
+	public SRPolicyCandidatePathIdentifiersTLV(){
+		this.TLVType=57;
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java:64]: change 57 to Objectparameters.xxxx (Modify ObjectParameters class)

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+   specified in [I-D.ietf-spring-segment-routing-policy] Section 2.4.
+
+   Discriminator: 32-bit value that encodes the Discriminator of the
+   candidate path.
+   
+	 */
+	
+	private int protocol;
+	
+	private int reserved;
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java:55]: Remove reserved. Just ignore the bytes upon reception

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+		return protocol;
+	}
+
+	public void setProtocol(int protocol) {
+		this.protocol = protocol;
+	}
+	
+	
+
+	public int getReserved() {
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java:82]: remove reserved

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+		return reserved;
+	}
+
+	public void setReserved(int reserved) {
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java:86]: remove reserved

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+	}
+
+	@Override
+	public void encode() {
+		//Log.debug("Encoding SRPolicyCandidatePathIdentifiers TLV");
+		this.setTLVValueLength(16);
+		this.tlv_bytes=new byte[this.getTotalTLVLength()];
+		this.encodeHeader();
+		int offset=4;
+		this.reserved=0;
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java:151]: Remove reserved

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+		int offset=4;//Position of the next subobject
+		if (this.getTLVValueLength()==0)
+		{
+			throw new MalformedPCEPObjectException();
+		}
+			
+		try {
+			this.protocol = ByteHandler.decode1byteInteger(this.tlv_bytes, offset);
+			offset+=1;
+			this.reserved = ByteHandler.decode3bytesInteger(this.tlv_bytes, offset);
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java:178]: No need to read the reserved bytes, just update the offset.

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+		int offset=4;
+		this.reserved=0;
+		
+		ByteHandler.encode1byteInteger(protocol, this.tlv_bytes, offset);
+		offset+=1;
+		ByteHandler.encode3bytesInteger(reserved, this.tlv_bytes, offset);
+		offset+=3;
+		ByteHandler.encode4bytesLong(originatorASN, this.tlv_bytes, offset);
+		offset+=4;
+		System.arraycopy(originatorAddress.getAddress(),0,this.tlv_bytes,offset,4);
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java:159]: Be careful,  Originator Address: Represented as 128 bit value where IPv4 address
   are encoded in lowest 32 bits, part of the originator identifier, as
   specified in [I-D.ietf-spring-segment-routing-policy] Section 2.4.
You are only econding the 32 bits of the IPv4 address....
 in [I-D.ietf-spring-segment-routing-policy] Section 2.4.:
 Node Address : represented as a 128-bit value.  IPv4 addresses
      MUST be encoded in the lowest 32 bits, and the high-order bits
      MUST be set to zero.
So... set all High order bits to 0 (that is upper 96 bits, about 10 bytes of Zeros)

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+			
+		try {
+			this.protocol = ByteHandler.decode1byteInteger(this.tlv_bytes, offset);
+			offset+=1;
+			this.reserved = ByteHandler.decode3bytesInteger(this.tlv_bytes, offset);
+			offset+=3;
+			this.originatorASN = ByteHandler.decode4bytesLong(this.tlv_bytes, offset);
+			offset+=4;
+			byte[] ip=new byte[4]; 
+			System.arraycopy(this.tlv_bytes,offset, ip, 0, 4);
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java:183]: Be careful when changing to send 128 bits for the originator. By now ignore all the 96 upper bits...

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathNameTLV.java [hunk 17]
```diff
@@ -0,0 +1,103 @@
+package es.tid.pce.pcep.objects.tlvs;
+
+import java.util.Objects;
+
+import es.tid.pce.pcep.objects.MalformedPCEPObjectException;
+import es.tid.pce.pcep.objects.ObjectParameters;
+
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathNameTLV.java:7]: add header with name, type, short description, reference and author

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathNameTLV.java [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+		byte[] SRPolicyCandidatePathNamePaso = new byte[this.getTLVValueLength()];
+		int offset=4;//Position of the next subobject
+		if (this.getTLVValueLength()==0)
+		{
+			throw new MalformedPCEPObjectException();
+		}
+		try {
+			System.arraycopy(this.tlv_bytes, offset, SRPolicyCandidatePathNamePaso, 0, this.getTLVValueLength());
+			this.SRPolicyCandidatePathName = new String(SRPolicyCandidatePathNamePaso);
+			log.debug("Sended name : "+this.SRPolicyCandidatePathName);
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathNameTLV.java:58]:
```suggestion
			log.debug("Candidate name : "+this.SRPolicyCandidatePathName);
```

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyName.java [hunk 19]
```diff
@@ -0,0 +1,119 @@
+package es.tid.pce.pcep.objects.tlvs;
+
+import java.util.Objects;
+
+import es.tid.pce.pcep.objects.MalformedPCEPObjectException;
+import es.tid.pce.pcep.objects.ObjectParameters;
+
+/*
+ * SR Policy Name TLV
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyName.java:9]: Add TVL Type code, a short description and the reference

File: src/main/java/es/tid/pce/pcep/objects/AssociationIPv6.java [hunk 20]
```diff
@@ -220,6 +267,18 @@ public void decode() throws MalformedPCEPObjectException {
 			case ObjectParameters.PCEP_TLV_EXTENDED_ASSOCIATION_ID:
 				extended_ssociation_id_tlv=new ExtendedAssociationIDTLV(this.getObject_bytes(), offset);
 				break;
+			case 56:
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/AssociationIPv6.java:270]: Change to case ObjectParameters.xxx

File: src/main/java/es/tid/pce/pcep/objects/AssociationIPv6.java [hunk 21]
```diff
@@ -220,6 +267,18 @@ public void decode() throws MalformedPCEPObjectException {
 			case ObjectParameters.PCEP_TLV_EXTENDED_ASSOCIATION_ID:
 				extended_ssociation_id_tlv=new ExtendedAssociationIDTLV(this.getObject_bytes(), offset);
 				break;
+			case 56:
+				sr_policy_name = new SRPolicyName(this.getObject_bytes(),offset);
+				break;
+			case 57:
+				sr_policy_candidate_path_identifiers_tlv= new SRPolicyCandidatePathIdentifiersTLV(this.getObject_bytes(),offset);
+				break;
+			case 58:
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/AssociationIPv6.java:276]: Change to case ObjectParameters.xxx

File: src/main/java/es/tid/pce/pcep/objects/AssociationIPv6.java [hunk 22]
```diff
@@ -220,6 +267,18 @@ public void decode() throws MalformedPCEPObjectException {
...
+			case 56:
+				sr_policy_name = new SRPolicyName(this.getObject_bytes(),offset);
+				break;
+			case 57:
+				sr_policy_candidate_path_identifiers_tlv= new SRPolicyCandidatePathIdentifiersTLV(this.getObject_bytes(),offset);
+				break;
+			case 58:
+				sr_policy_candidate_path_tlv=new SRPolicyCandidatePathNameTLV(this.getObject_bytes(),offset);
+				break;
+			case 59:
```
Comment (oscargdd) [inline src/main/java/es/tid/pce/pcep/objects/AssociationIPv6.java:279]: Change to case ObjectParameters.xxx

File: src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java (hunk 5 above)
Comment (Luis-Cepeda) [reply src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java]: Done!

File: src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java (hunk 7 above)
Comment (Luis-Cepeda) [reply src/main/java/es/tid/pce/pcep/objects/AssociationIPv4.java]: Done!

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java (hunk 9 above)
Comment (Luis-Cepeda) [reply src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java]: Done!

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java (hunk 14 above)
Comment (Luis-Cepeda) [reply src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyCandidatePathIdentifiersTLV.java]: Done!

File: src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyName.java (hunk 19 above)
Comment (Luis-Cepeda) [reply src/main/java/es/tid/pce/pcep/objects/tlvs/SRPolicyName.java:9]: Done!

File: src/main/java/es/tid/pce/pcep/objects/AssociationIPv6.java (hunk 22 above)
Comment (Luis-Cepeda) [reply src/main/java/es/tid/pce/pcep/objects/AssociationIPv6.java]: Done!

**Pull request merged**